### ISSUE TITLE
[BUG] Rancher server startup gets stuck if KDM is blocked

### ISSUE THREAD

**Issue by niusmallnan** [rancher/rancher#40416]
**Rancher Server Setup**
- Rancher version: v2.7.1/v2.6.11
- Installation option (Docker install/Helm Chart):
   - If Helm Chart, Kubernetes Cluster and version (RKE1, RKE2, k3s, EKS, etc): Any
- Proxy/Cert Details:

**Describe the bug**

KDM means `https://github.com/rancher/kontainer-driver-metadata` .
KDM uses Cloudfront, and in certain areas, CloudFront IPs may be blocked by firewalls. A special blocking method, KDM will not be denied the request, it has allowed the connection but cannot read the data normally.

When the Rancher server starts, it will try to read the remote address of KDM (https://releases.rancher.com/kontainer-driver-metadata/release-v2.7/data.json). Since the code is not perfect, the request timeout is not set, which would prevent the request from exiting properly.

Usually, the blocking of a request does not affect the overall startup process. Unfortunately, we put this request process in `sync.Once`. Then, the Rancher server will always be stuck and unable to complete the startup.

I fetch the Goroutine stack:
```
goroutine 1 [sync.Cond.Wait]:
sync.runtime_notifyListWait(0xc002699248, 0x52)
	/usr/lib64/go/1.19/src/runtime/sema.go:517 +0x14c
sync.(*Cond).Wait(0xd?)
	/usr/lib64/go/1.19/src/sync/cond.go:70 +0x8c
net/http.(*http2pipe).Read(0xc002699230, {0xc002d56c48, 0x413b8, 0x413b8})
	/usr/lib64/go/1.19/src/net/http/h2_bundle.go:3668 +0xeb
net/http.http2transportResponseBody.Read({0xc0020ae650?}, {0xc002d56c48?, 0x7f1523f895f8?, 0x7f15240163f8?})
	/usr/lib64/go/1.19/src/net/http/h2_bundle.go:9162 +0x85
io.ReadAll({0x7f1523f895f8, 0xc002699200})
	/usr/lib64/go/1.19/src/io/io.go:661 +0xfe
io/ioutil.ReadAll(...)
	/usr/lib64/go/1.19/src/io/ioutil/ioutil.go:27
github.com/rancher/channelserver/pkg/config.get({0x484c508, 0xc000912200}, {0x4819040, 0x6d10098})
	/go/pkg/mod/github.com/rancher/channelserver@v0.5.1-0.20220405170618-28c9b37deff1/pkg/config/get.go:54 +0x211
github.com/rancher/channelserver/pkg/config.getURLs({0x484c508, 0xc000912200}, {0xc00002af20?, 0x2, 0x40d9bf?})
	/go/pkg/mod/github.com/rancher/channelserver@v0.5.1-0.20220405170618-28c9b37deff1/pkg/config/get.go:24 +0x99
github.com/rancher/channelserver/pkg/config.(*Config).loadConfig(0x7f152424f598?, {0x484c508, 0xc000912200}, {0x3f35470, 0x3}, {0xc0000800b6, 0x6}, {0x3f41477, 0x7}, {0xc00002af20, ...})
	/go/pkg/mod/github.com/rancher/channelserver@v0.5.1-0.20220405170618-28c9b37deff1/pkg/config/config.go:77 +0x91
github.com/rancher/channelserver/pkg/config.NewConfig({0x484c508?, 0xc000912200}, {0x3f35470, 0x3}, {0x4819020?, 0xc0004fa920}, {0xc0000800b6, 0x6}, {0x3f41477, 0x7}, ...)
	/go/pkg/mod/github.com/rancher/channelserver@v0.5.1-0.20220405170618-28c9b37deff1/pkg/config/config.go:60 +0x1df
github.com/rancher/rancher/pkg/channelserver.GetReleaseConfigByRuntime.func1()
	/go/src/github.com/rancher/rancher/pkg/channelserver/channelserver.go:120 +0x113
sync.(*Once).doSlow(0x405a45?, 0x20?)
	/usr/lib64/go/1.19/src/sync/once.go:74 +0xc2
sync.(*Once).Do(...)
	/usr/lib64/go/1.19/src/sync/once.go:65
github.com/rancher/rancher/pkg/channelserver.GetReleaseConfigByRuntime({0x484c508?, 0xc000912200?}, {0x3f35470?, 0x3?})
	/go/src/github.com/rancher/rancher/pkg/channelserver/channelserver.go:114 +0x71
github.com/rancher/rancher/pkg/channelserver.NewHandler({0x484c508, 0xc000912200})
	/go/src/github.com/rancher/rancher/pkg/channelserver/channelserver.go:130 +0xca
github.com/rancher/rancher/pkg/multiclustermanager.router({0x484c508, 0xc000912200}, 0x60?, 0xc00167f9c8?, 0xc001c50700, 0xc0019048c0)
	/go/src/github.com/rancher/rancher/pkg/multiclustermanager/routes.go:70 +0x37a
github.com/rancher/rancher/pkg/multiclustermanager.newMCM({0x484c508?, 0xc000912200}, 0xc001ac5860, 0xc00167f9c8)
	/go/src/github.com/rancher/rancher/pkg/multiclustermanager/app.go:110 +0x7a
github.com/rancher/rancher/pkg/multiclustermanager.(*DeferredServer).Start.func1({0x484c508, 0xc000912200})
	/go/src/github.com/rancher/rancher/pkg/multiclustermanager/deferred.go:75 +0x45
github.com/rancher/rancher/pkg/wrangler.(*Context).StartWithTransaction(0xc001ac5860, {0x484beb0?, 0xc0019c6b80}, 0xc0015ecf58)
	/go/src/github.com/rancher/rancher/pkg/wrangler/context.go:154 +0xdf
github.com/rancher/rancher/pkg/multiclustermanager.(*DeferredServer).Start(0xc001c53e00, {0x484beb0?, 0xc0019c6b80})
	/go/src/github.com/rancher/rancher/pkg/multiclustermanager/deferred.go:74 +0x105
github.com/rancher/rancher/pkg/rancher.(*Rancher).Start(0xc0019c6f40, {0x484beb0, 0xc0019c6b80})
	/go/src/github.com/rancher/rancher/pkg/rancher/rancher.go:242 +0xa2
github.com/rancher/rancher/pkg/rancher.(*Rancher).ListenAndServe(0xc0019c6f40, {0x484beb0, 0xc0019c6b80})
	/go/src/github.com/rancher/rancher/pkg/rancher/rancher.go:270 +0x35
main.run(0x12?, {{0x0, 0x0, 0x0}, {0x3f36cf6, 0x4}, 0x0, {0x0, 0x0}, 0x50, ...})
	/go/src/github.com/rancher/rancher/main.go:227 +0x305
main.main.func1(0xc0019be9a0?)
	/go/src/github.com/rancher/rancher/main.go:173 +0xe5
github.com/urfave/cli.HandleAction({0x35600a0?, 0xc00199afd0?}, 0xc0016aee00?)
	/go/pkg/mod/github.com/urfave/cli@v1.22.9/app.go:524 +0x50
github.com/urfave/cli.(*App).Run(0xc0016aee00, {0xc000072080, 0x8, 0x8})
	/go/pkg/mod/github.com/urfave/cli@v1.22.9/app.go:286 +0x7ca
main.main()
	/go/src/github.com/rancher/rancher/main.go:180 +0xfcc
```

**To Reproduce**

Use the following code to simulate Slow Response HTTP:

```
import http.server
import socketserver
from time import sleep

PORT = 5000
SLEEP_TIME = 600

class SlowHandler(http.server.SimpleHTTPRequestHandler):
    def do_GET(self):
        sleep(SLEEP_TIME)
        http.server.SimpleHTTPRequestHandler.do_GET(self)

Handler = SlowHandler
httpd = socketserver.TCPServer(("", PORT), Handler)

print("serving at port", PORT)
httpd.serve_forever()
```
Run it with Python3 and point KDM to it when starting Rancher:
```
docker run -d --restart=unless-stopped \
    -e CATTLE_RKE_METADATA_CONFIG="{\"url\":\"http://10.0.7.8:5000\", \"refresh-interval-minutes\":\"0\"}" \
    -e CATTLE_BOOTSTRAP_PASSWORD=RancherForFun \
    -p 80:80 -p 443:443 \
    --privileged \
    rancher/rancher:v2.7.1
```

Checking the logs, we can see when the Rancher server gets stuck:
```
2023/02/03 14:03:20 [INFO] Done waiting for CRD cisbenchmarkversions.management.cattle.io to become available
2023/02/03 14:23:21 [INFO] Starting management.cattle.io/v3, Kind=ClusterRoleTemplateBinding controller
```

The root cause of this problem is that the appropriate HTTP request timeout is not set in the channelserver lib.
https://github.com/rancher/channelserver/blob/cdef874cc406bdffbd67abe8067e0f76e950c20a/pkg/config/get.go#L33-L55

**Comment by jiaqiluo** [comment]
PR was merged, waiting for rancher build to pass

**Comment by jiaqiluo** [comment]
This issue can be validated on  v2.7-head 

The steps are in the description of the issue.

**Comment by rishabhmsra** [comment]
Verified on rancher v2.7-head(c00569d)

Steps performed:
- Created a VM on AWS and started the slow response http server.
- Started rancher v2.7-head with `CATTLE_RKE_METADATA_CONFIG` pointing to the http server.
- Observed that logs did not get stuck and rancher server was up and running.

Also verifed on rancher v2.7.1
- Performed same steps as above and observed that logs get stuck at:

```
[INFO] Waiting for CRD templates.management.cattle.io to become available
[INFO] Done waiting for CRD templates.management.cattle.io to become available
```
Result:
- rancher v2.7-head starting up with slow response http server and logs are not getting stuck, hence closing the issue.
